astroid/inference_tip.py

```diff
@@ -35,12 +34,18 @@ def clear_inference_tip_cache() -> None:
 
 def _inference_tip_cached(
     func: Callable[_P, Iterator[InferenceResult]],
-) -> Callable[_P, Iterator[InferenceResult]]:
+) -> InferFnExplicit:
     """Cache decorator used for inference tips."""
 
-    def inner(*args: _P.args, **kwargs: _P.kwargs) -> Iterator[InferenceResult]:
+    def inner(
+        *args: _P.args, **kwargs: _P.kwargs
+    ) -> Iterator[InferenceResult] | list[InferenceResult]:
         node = args[0]
         context = args[1]
+        if TYPE_CHECKING:
+            assert isinstance(node, NodeNG)
+            assert context is None or isinstance(context, InferenceContext)
```

Comment:
```suggestion
        node: nodeNG = args[0]
        context: InferenceContext | None = args[1]
```

Think I would prefer this?

Reply:
Actually, this doesn't work, gives:
```
astroid/inference_tip.py:44: error: Incompatible types in assignment (expression has type "object", variable has type "NodeNG")  [assignment]
astroid/inference_tip.py:45: error: Incompatible types in assignment (expression has type "object", variable has type "Optional[InferenceContext]")  [assignment]
```
I'll cast instead of asserting. I wanted to avoid the penalty in such a hot path of isinstance.

Reply:
Let's add a type ignore. If we improve the typing this might at some point then be flagged as unused.

Reply:
[5fd854f](https://github.com/pylint-dev/astroid/pull/2166/commits/5fd854fff4439c4b322a706687d069e25e6a802e)

Reply:
> If we improve the typing this might at some point then be flagged as unused.

How could the typing be better, unless we remove the dynamic *args? (An earlier version of this PR did that, but I wasn't sure that was alright to remove support for it.)

Reply:
I would say we can remove support, it's breaking changes season anyway.

Reply:
⛈️

Reply:
06f42bb